Add --relocatable support for tables

wingo opened the pull request:
This patch adds a table symbol type and plumbs it through everywhere.

To do:

- [x] Add tests
- [x] Determine what the right thing is for modules that implicitly use table 0 but don't declare a table (see https://github.com/WebAssembly/wabt/compare/master...wingo:relocatable-tables?expand=1#diff-4803dd58b37da26b663dfaa7d34ba35e)
  - The answer here is a bit more straightforward: the initial patch wasn't emitting symtab entries for anonymous definitions.  The test case under consideration did in fact define the table.  The updated patch will emit relocs for all table number references, unless the reference is invalid/broken (the user must have passed --no-check).
- [x] Finalize the reloc kind for `R_WASM_TABLE_NUMBER_LEB` in [Linking.md](https://github.com/WebAssembly/tool-conventions/blob/master/Linking.md); @pmatos is on this.
- [x] Only emit relocs when reftypes are enabled (see https://github.com/WebAssembly/wabt/pull/1549#issuecomment-697480009)

Cc @pmatos and @sbc100

sbc100 commented:
This generally looks great!

Personally I'd rather see these changes land in llvm / wasm-ld before we land them in wabt since as far as I know wasm-ld is the only consumer of these relocations right now.  Are you planning a corresponding llvm change?

Perhaps we can discuss this more on Monday with @pmatos.. but I was wondering if we should handle table 0 and memory 0 specially.   In llvm memory 0 is currently special in that it represents the linear memory of the program and table 0 is special in that it holds all the indirect function addresses.   I could imagine world where we could avoid relocations if the index is 0, avoiding the need for all call_indirect instructions to have an extra relocation, but I'm not set on this.

pmatos commented:
> This generally looks great!
> 
> Personally I'd rather see these changes land in llvm / wasm-ld before we land them in wabt since as far as I know wasm-ld is the only consumer of these relocations right now. Are you planning a corresponding llvm change?
> 

Yes - I plan to work on this as soon as I finish a couple of other small tasks.

> Perhaps we can discuss this more on Monday with @pmatos.. but I was wondering if we should handle table 0 and memory 0 specially. In llvm memory 0 is currently special in that it represents the linear memory of the program and table 0 is special in that it holds all the indirect function addresses. I could imagine world where we could avoid relocations if the index is 0, avoiding the need for all call_indirect instructions to have an extra relocation, but I'm not set on this.

Lets have a chat about this on Monday as planned so we can lay the ground for future work.

wingo commented:
> This generally looks great!

Thanks for taking a look!

> Personally I'd rather see these changes land in llvm / wasm-ld before we land them in wabt since as far as I know wasm-ld is the only consumer of these relocations right now. Are you planning a corresponding llvm change?

Agreed, and hopefully we can get that patch up this week too.

> I was wondering if we should handle table 0 and memory 0 specially. In llvm memory 0 is currently special in that it represents the linear memory of the program and table 0 is special in that it holds all the indirect function addresses. I could imagine world where we could avoid relocations if the index is 0, avoiding the need for all call_indirect instructions to have an extra relocation, but I'm not set on this.

I was wondering about this too.  Concretely, looking at the compilation of:

```c
void f(void) {}
void* g(void) { return f; }
```

I see that the current result objdumps as:

```
Type[2]:
 - type[0] () -> nil
 - type[1] () -> i32
Import[2]:
 - memory[0] pages: initial=0 <- env.__linear_memory
 - table[0] type=funcref initial=1 <- env.__indirect_function_table
Function[2]:
 - func[0] sig=0 <f>
 - func[1] sig=1 <g>
Elem[1]:
 - segment[0] flags=0 table=0 count=1 - init i32=1
  - elem[1] = func[0] <f>
Code[2]:
 - func[0] size=2 <f>
 - func[1] size=8 <g>
Custom:
 - name: "linking"
  - symbol table [count=2]
   - 0: F <f> func=0 binding=global vis=hidden
   - 1: F <g> func=1 binding=global vis=hidden
Custom:
 - name: "reloc.CODE"
  - relocations for section: 4 (Code) [1]
   - R_WASM_TABLE_INDEX_SLEB offset=0x000007(file=0x000079) symbol=0 <f>
Custom:
 - name: "producers"

Code Disassembly:

000074 func[0] <f>:
 000075: 0b                         | end
000077 func[1] <g>:
 000078: 41 81 80 80 80 00          | i32.const 1
 00007e: 0b                         | end
```

If we want to omit relocs for the indirect function table, we can do so if the referred-to table has index 0, is imported, has type `funcref`, and the import is from the `env` module, and the name of the import is `__indirect_function_table`.

The advantage of doing so is that you would have fewer relocs to deal with, which probably don't need to be renumbered from 0 in the "normal" case, and for which you wouldn't have to reserve a 5-byte LEB.

The disadvantage is that the filter logic described above is a bit of a hack.  Also I would imagine that in the future as CFI techniques start to be more standard, and with the advent of typed function references, we'll start to see more partitioning of callables into multiple tables, so this optimization is not one for the long term, IMO.

For me, I think I would lean towards adding `R_WASM_TABLE_NUMBER_LEB` relocs for all table number references.  If short-term efficiency is still a consideration, we can only emit such relocs when reftypes are enabled, in wabt anyway; not sure how feature selection works in LLVM yet.

sbc100 commented:
> > This generally looks great!
> 
> Thanks for taking a look!
> 
> > Personally I'd rather see these changes land in llvm / wasm-ld before we land them in wabt since as far as I know wasm-ld is the only consumer of these relocations right now. Are you planning a corresponding llvm change?
> 
> Agreed, and hopefully we can get that patch up this week too.
> 
> > I was wondering if we should handle table 0 and memory 0 specially. In llvm memory 0 is currently special in that it represents the linear memory of the program and table 0 is special in that it holds all the indirect function addresses. I could imagine world where we could avoid relocations if the index is 0, avoiding the need for all call_indirect instructions to have an extra relocation, but I'm not set on this.
> 
> I was wondering about this too. Concretely, looking at the compilation of:
> 
> ```c
> void f(void) {}
> void* g(void) { return f; }
> ```
> 
> I see that the current result objdumps as:
> 
> ```
> Type[2]:
>  - type[0] () -> nil
>  - type[1] () -> i32
> Import[2]:
>  - memory[0] pages: initial=0 <- env.__linear_memory
>  - table[0] type=funcref initial=1 <- env.__indirect_function_table
> Function[2]:
>  - func[0] sig=0 <f>
>  - func[1] sig=1 <g>
> Elem[1]:
>  - segment[0] flags=0 table=0 count=1 - init i32=1
>   - elem[1] = func[0] <f>
> Code[2]:
>  - func[0] size=2 <f>
>  - func[1] size=8 <g>
> Custom:
>  - name: "linking"
>   - symbol table [count=2]
>    - 0: F <f> func=0 binding=global vis=hidden
>    - 1: F <g> func=1 binding=global vis=hidden
> Custom:
>  - name: "reloc.CODE"
>   - relocations for section: 4 (Code) [1]
>    - R_WASM_TABLE_INDEX_SLEB offset=0x000007(file=0x000079) symbol=0 <f>
> Custom:
>  - name: "producers"
> 
> Code Disassembly:
> 
> 000074 func[0] <f>:
>  000075: 0b                         | end
> 000077 func[1] <g>:
>  000078: 41 81 80 80 80 00          | i32.const 1
>  00007e: 0b                         | end
> ```
> 
> If we want to omit relocs for the indirect function table, we can do so if the referred-to table has index 0, is imported, has type `funcref`, and the import is from the `env` module, and the name of the import is `__indirect_function_table`.
> 
> The advantage of doing so is that you would have fewer relocs to deal with, which probably don't need to be renumbered from 0 in the "normal" case, and for which you wouldn't have to reserve a 5-byte LEB.
> 
> The disadvantage is that the filter logic described above is a bit of a hack. Also I would imagine that in the future as CFI techniques start to be more standard, and with the advent of typed function references, we'll start to see more partitioning of callables into multiple tables, so this optimization is not one for the long term, IMO.
> 
> For me, I think I would lean towards adding `R_WASM_TABLE_NUMBER_LEB` relocs for all table number references. If short-term efficiency is still a consideration, we can only emit such relocs when reftypes are enabled, in wabt anyway; not sure how feature selection works in LLVM yet.

I like this last idea.  I think I agree that for consistency we should probably use relocations everywhere and that enabling this only when reftypes are enabled is a good way to transition to this new world.  

Also we look at adding `R_WASM_MEMORY_NUMBER_LEB` at the same time support multi-memory, or maybe this can wait until we have more concrete users of the multi memory proposal?

wingo commented:
Thanks for the review!

Regarding multi-memory, it's enough boilerplate that it probably warrants a separate PR.  I filed #1552.

I updated the TODO at the top and added an entry for only emitting table relocs if reftypes are enabled.  I also added a basic test.

pmatos commented:
>     * [ ]  Finalize the reloc kind for `R_WASM_TABLE_NUMBER_LEB` in [Linking.md](https://github.com/WebAssembly/tool-conventions/blob/master/Linking.md); @pmatos is on this.
> 

See https://github.com/WebAssembly/tool-conventions/pull/147

binji reviewed (approved):
lgtm

wingo commented:
Ready to merge, I think.  I know that wasm-ld work is ongoing but there is also a bug-fix in this patch; the logic at https://github.com/WebAssembly/wabt/pull/1549/files#diff-f69899e3f8be9636bcd8d71e903ea8a0R238 wasn't right before.  (All function references need relocs, so all functions should be in the symbol table, regardless of whether they have a name or not.)  If wabt needs followup fixes for the table_number reloc, I will make sure to take care of it.

sbc100 commented on src/binary-reader-objdump.cc:
```diff
@@ -1781,11 +1809,23 @@ Result BinaryReaderObjdump::OnEventSymbol(Index index,
     name = GetEventName(event_index);
   }
-  PrintDetails("   - [%d] E <" PRIstringview "> event=%" PRIindex, index,
+  PrintDetails("   - %d: E <" PRIstringview "> event=%" PRIindex, index,
```
Looks like we don't have any test that tickles this?  (Since I don't see any corresponding test change due to this?)

wingo replied:
Indeed -- just noticed this as I had copied the table implementation from events, and events were printing inconsistently with regards to everything else, so I just fixed this one up.  Indeed, apparently no tests.

sbc100 commented:
Oh .. we like to do a single commit per PR in this repo.   It looks like you have diligently split you your into commits with nice separate messages.

Would you mind creating a single message that you would like to use for the final commit message?  You could edit the PR description to include this and I can copy and paste it into the squashed commit?

wingo commented:
@sbc100 I squashed, rebased, and updated the commit message, though I see now that you were just asking for an updated message and that you would do the squashing :grimacing: sorry about that.  In any case it's the same code, rebased cleanly onto today's master -- no manual edits.
